### PR TITLE
feat: SDK defaults add disable profiler to createTrainingJob

### PR DESCRIPTION
**Pull request by akrishna1995**
added profilerConfig to the traningjob schema
for sagemaker_config

disableprofiler attribute is injected via config
in session,estimator

*Issue #, if available:*

*Description of changes:*

*Testing done:*

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [x] I have read the [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md) doc
- [x] I certify that the changes I am introducing will be backward compatible, and I have discussed concerns about this, if any, with the Python SDK team
- [x] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md#committing-your-change)
- [ ] I have passed the region in to all S3 and STS clients that I've initialized as part of this change.
- [ ] I have updated any necessary documentation, including [READMEs](https://github.com/aws/sagemaker-python-sdk/blob/master/README.rst) and [API docs](https://github.com/aws/sagemaker-python-sdk/tree/master/doc) (if appropriate)

#### Tests

- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [x] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes
- [ ] I have checked that my tests are not configured for a specific region or account (if appropriate)
- [ ] I have used [`unique_name_from_base`](https://github.com/aws/sagemaker-python-sdk/blob/master/src/sagemaker/utils.py#L77) to create resource names in integ tests (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by akrishna1995 — COMMENTED** [review]
/bot run slow-tests

**Review by akrishna1995 — COMMENTED** [review]
/bot run all

File: src/sagemaker/estimator.py
Comment (knikure) [inline src/sagemaker/estimator.py:689]: `default_value` for `disable_profiler` is False.

File: src/sagemaker/config/config_schema.py
Comment (knikure) [inline src/sagemaker/config/config_schema.py:135]: Can we reuse the TRAINING_JOB_PROFILE_CONFIG_PATH here? 
So it can be TRAINING_JOB_PROFILE_CONFIG_PATH, DISABLE_PROFILER.

File: src/sagemaker/estimator.py
Comment (rubanh) [reply src/sagemaker/estimator.py:689]: Agreed. And be sure to change the constructor also, from 

```
        disable_profiler: bool = False,
```
into:
```
        disable_profiler: bool = None,
```

File: src/sagemaker/estimator.py
Comment (rubanh) [inline src/sagemaker/estimator.py:685]: Seems profiler_config is a class sagemaker.debugger.ProfilerConfig ?

So pulling the dict from the config and setting it directly might lead to unexpected behavior. Do we need this injection anyway? Later as more params are added to the profiler_config, I think the existing class helper could be used

**Review by rubanh — CHANGES_REQUESTED** [review]
overall looks good! just some minor changes needed

File: src/sagemaker/config/config_schema.py
Comment (akrishna1995) [reply src/sagemaker/config/config_schema.py:135]: neat will make the update!

File: src/sagemaker/estimator.py
Comment (akrishna1995) [reply src/sagemaker/estimator.py:685]: Makes sense. Great callout - I can kick it out and only inject the disable_profiler value in.

Comment (knikure) [reply src/sagemaker/estimator.py:685]: `resolve_class_attribute_from_config` is the right choice here.

File: src/sagemaker/session.py
Comment (knikure) [inline src/sagemaker/session.py:832]: Same. Use `resolve_class_attribute_from_config` here as well.

File: src/sagemaker/estimator.py
Comment (knikure) [inline src/sagemaker/estimator.py:41]: Can we just name this `TRAINING_JOB_DISABLE_PROFILER_PATH` as per the existing naming convention followed?

File: tests/unit/test_session.py
Comment (knikure) [inline tests/unit/test_session.py:1702]: Curious, what happens when `DisableProfiler` is set to `True` and `profiler_config` is provided though function input?

File: tests/unit/test_session.py
Comment (knikure) [inline tests/unit/test_session.py:1666]: What is the default for `profiler_config` when nothing is provided? We know its None when we pass params to train() function but later it does do any defaults by setting some dict value? Worth checking that as well.

File: src/sagemaker/session.py
Comment (akrishna1995) [reply src/sagemaker/session.py:832]: hmmm in estimator.py the profiler_config is of type ```ProfilerConfig``` , but in session.py the doc mentions it is just a dict ? Would we still need `resolve_class_attirbute_from_config` ?

File: tests/unit/test_session.py
Comment (akrishna1995) [reply tests/unit/test_session.py:1702]: The behavior of `update_nested_dictionary_with_values_from_config` is - if the profiler_config is populated with the disableProfiler attribute then that will be picked over what is present in the SDK defaults. if it is partially updated with out the value the one in the config will be used, similarly None or empty

I can write a UT to validate this as well

File: tests/unit/test_session.py
Comment (akrishna1995) [reply tests/unit/test_session.py:1666]: Makes sense, I checked session.py and did not see any function implementation of the profiler_config object with function defaults - great callout and learnings from the DebugHookConfig. 

Let me know if there are any other places to check though

File: src/sagemaker/session.py
Comment (akrishna1995) [reply src/sagemaker/session.py:832]: sycned offline - weird impl, can keep it as dict

File: src/sagemaker/estimator.py
Comment (rubanh) [inline src/sagemaker/estimator.py:359]: Ah so this we want to keep as False :D 
Because if the user passes in nothing, and nothing is set in the config, the constructor will still set it to False, (but only after checking the config, whereas before it was done thru the constructor default value)

File: tests/unit/test_session.py
Comment (rubanh) [reply tests/unit/test_session.py:1732]: since disable_profiler is set the False by default, the SDK should have some default profiler config it uses. But that might be before the session.py is called.

Mentioning and reopening this comment in case its relevant. Can close it again if its not relevant

File: tests/unit/test_session.py
Comment (rubanh) [inline tests/unit/test_session.py:1627]: optional: might be able to parameterize these similar tests for cleaner code. Could reference some of the sagemaker_config util tests to see how to do parameterization like that

File: tests/unit/test_session.py
Comment (rubanh) [inline tests/unit/test_session.py:1834]: would be nice to also make sure the other profiler config fields from the user's input weren't erased or anything

**Review by rubanh — APPROVED** [review]
One small docs change that should be made before merging (dont need another approval from me). Other changes are optional

File: src/sagemaker/session.py
Comment (knikure) [inline src/sagemaker/session.py:873]: Is the a reason we are not supporting this in update_training_job() function? Refer Line 1152.

File: src/sagemaker/estimator.py
Comment (akrishna1995) [reply src/sagemaker/estimator.py:359]: Ahhhh Makes sense - Sorry misunderstood your previous comment - Thanks for the catch :)

File: src/sagemaker/session.py
Comment (akrishna1995) [reply src/sagemaker/session.py:873]: I believe update_training_job is out of scope, we need to support **disable_profiler**  for CreateTrainingJob, in Estimator and Session according to the customer requirements doc but can double check

Comment (akrishna1995) [reply src/sagemaker/session.py:873]: Discussed offline: We will incorporate this using the same config injection for CreateTrainingJob schema and update in next commit

File: tests/unit/test_session.py
Comment (akrishna1995) [reply tests/unit/test_session.py:1732]: I double checked the session.py and confirmed that there is no custom dict being built, for estimator.py we inject only the disable_profiler value. I think we should be good - Thanks for the callout :)

File: tests/unit/test_session.py
Comment (akrishna1995) [reply tests/unit/test_session.py:1627]: Will optimize as a follow up in another PR

File: src/sagemaker/session.py
Comment (knikure) [inline src/sagemaker/session.py:1143]: Add a unit test to cover update_training_job() functional scenarios.
You can also do that in a follow-up PR.

**Review by knikure — APPROVED** [review]
Approving with a follow-up item for unit-test.

**Pull request merged**
